scarlet/observation.py

```diff
@@ -105,74 +124,64 @@ class Observation(Scene):
+            self.structure = np.array(self.structure)
+        self.padding = padding
 
     def match(self, scene):
+        """Match the psf in each observed band to the target PSF
+        """
+        if self.psfs is not None:
+            # First we setup the parameters for the model -> observation FFTs
+            # Make the PSF stamp wider due to errors when matching PSFs
+            psf_shape = np.array(self.psfs[0].shape) + self.padding
+            shape = np.array(scene.shape[1:]) + psf_shape - 1
+            # Choose the optimal shape for FFTPack DFT
+            self.fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape]
+            # Store the pre-fftpack optimization slices
+            self.slices = tuple([slice(s) for s in shape])
+
+            # Now we setup the parameters for the psf -> kernel FFTs
+            shape = np.array(scene.psfs[0].shape) + np.array(self.psfs[0].shape) - 1
+            fftpack_shape = [fftpack.helper.next_fast_len(d) for d in shape]
+            # Deconvolve the target PSF
+            target_fft = np.fft.rfftn(scene.psfs[0], fftpack_shape)
 
-        # 1) determine shape of scene in obs, set mask
-
-        # 2) compute the interpolation kernel between scene and obs
-
-        # 3) compute obs.psf in the frame of scene, store in Fourier space
-        # A few notes on this procedure:
-        # a) This assumes that scene.psfs and self.psfs have the same spatial shape,
-        #    which will need to be modified for multi-resolution datasets
-        if self._psfs is not None:
-            ipad, ppad = interpolation.get_common_padding(self.images, self._psfs, padding=self.padding)
-            self.image_padding, self.psf_padding = ipad, ppad
-            _psfs = np.pad(self._psfs, ((0, 0), *self.psf_padding), 'constant')
-            _target = np.pad(scene._psfs, self.psf_padding, 'constant')
-
+            # Match the PSF in each band
             new_kernel_fft = []
-            # Deconvolve the target PSF
-            target_fft = np.fft.fft2(np.fft.ifftshift(_target))
-
-            for _psf in _psfs:
-                observed_fft = np.fft.fft2(np.fft.ifftshift(_psf))
-                # Create the matching kernel
-                kernel_fft = observed_fft / target_fft
-                # Take the inverse Fourier transform to normalize the result
-                # Trials without this operation are slow to converge, but in the future
-                # we may be able to come up with a method to normalize in the Fourier Transform
-                # and avoid this step.
-                kernel = np.fft.ifft2(kernel_fft)
-                kernel = np.fft.fftshift(np.real(kernel))
-                kernel /= kernel.sum()
-                # Store the Fourier transform of the matching kernel
-                new_kernel_fft.append(np.fft.fft2(np.fft.ifftshift(kernel)))
+            kernels = []
+            for psf in self.psfs:
+                _psf_fft = np.fft.rfftn(psf, fftpack_shape)
```

Comment:
could we do the FFT with all bands at once and then split into the bands if needed?

Reply:
In fact, it looks like the entire `for` block could be done with all bands at once, which would also avoid the extra copies in lines 174-175

Reply:
This is not required work for this PR since it's #95

Reply:
I agree that this *should* be possible.